doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,433 @@
+   constructs.
+5. Before using any cxx container the user as to call `reserve()` once otherwise
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+The draft we provide here is using the `cxx::vector` but the techniques
+described can be easily applied to the `cxx::string` or other cxx containers.
+
+```
++---------------------------------------------------+
+|template<T>                                        |
+|vector                                             |
+|                                                   |
+|  - vector() = default                             |
+|  - template<Allocator>                            |
+|    vector(Allocator & allocator);                 |
+|                                                   |
+|  - void reserve(const uint64_t);                  |
+|  - void release();                                |
+|                                                   |
+|  // remaining API is unchanged                    |
++---------------------------------------------------+
+                          ^
+                         /|\
+                          |
+       +------------------+--------------------------+
+       |template<T, Capacity>                        |
+       |vector : public vector<Type>                 |
+       |                                             |
+       |  (#1)                                       |
+       |  - vector(const vector<T> &)                |
+       |  - vector(vector<T> &&)                     |
+       |  - vector& operator=(const vector<T> &)     |
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based version of cxx containers will not have constructors similar
+to the STL. The only provided constructor has one argument which specifies the
+type of allocator one would like to use for that container. If no argument is
+provided the heap allocator will be used by default.
+
+The constructors and operations defined in (#1) are added to support implicit
+conversion from the generic `cxx::vector` which is required to the stack version
+of a vector. It makes the following operations possible
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int> v1;
+vector<int, 20> v2;
+
+// this works thanks to implicit conversions from vector<int>
+f(v1);
+v2 = v1;
```

Comment:
What happens when the size of `v1` is greater than 20?

Reply:
It fails, a detailed explanation was added to the document.